File: chroma-league/src/main/java/net/booone/chromaleague/hud/parts/dragons/KilledDragonsBar.java
package net.booone.chromaleague.hud.parts.dragons;

import net.booone.chromaleague.rest.eventdata.DragonType;
import net.booone.chromaleague.state.GameStateHelper;
import net.booone.razersdk.animation.AnimatedFrame;
import net.booone.razersdk.animation.Frame;
import net.booone.razersdk.animation.SimpleFrame;
import net.booone.razersdk.color.Color;
import net.booone.razersdk.sdk.RzKey;

import java.util.*;

import static java.util.Collections.emptyMap;
import static net.booone.razersdk.sdk.RzKey.*;

public class KilledDragonsBar extends AnimatedFrame {
    public static final List<RzKey> FIRST_DRAGON_ROW = List.of(RZKEY_BACKSLASH);
    public static final List<RzKey> SECOND_DRAGON_ROW = List.of(RZKEY_ENTER);
    public static final List<RzKey> THIRD_DRAGON_ROW = List.of(RZKEY_RSHIFT);
    public static final List<RzKey> FOURTH_DRAGON_ROW = List.of(RZKEY_FN, RZKEY_RMENU, RZKEY_RCTRL);
    public static final List<RzKey> SOUL_BAR = List.of(RZKEY_LCTRL, RZKEY_LWIN, RZKEY_LALT, RZKEY_SPACE, RZKEY_RALT,
            RZKEY_UP, RZKEY_LEFT, RZKEY_DOWN, RZKEY_RIGHT);
    private boolean soulMode = false;

    private final Map<RzKey, Color> dragonColorsMap = new HashMap<>();

    @Override
    public Frame getFrame() {
        addAnimationFrame(new SimpleFrame(getKilledDragonsBar()));
        return super.getFrame();
    }

    private Map<RzKey, Color> getKilledDragonsBar() {
        Map<RzKey, Color> killedDragonsBar = new EnumMap<>(RzKey.class);
        Optional<DragonType> dragonSoul = getDragonSoul(GameStateHelper.getKilledDragons());
        if (dragonSoul.isPresent() && !soulMode) {
            DragonType soulType = dragonSoul.get();
            getDragonType(0).filter(dragonType -> dragonType == soulType).ifPresent(dragonType -> swapWithDragonSoulColor(soulType, FIRST_DRAGON_ROW));
            getDragonType(1).filter(dragonType -> dragonType == soulType).ifPresent(dragonType -> swapWithDragonSoulColor(soulType, SECOND_DRAGON_ROW));
            getDragonType(2).filter(dragonType -> dragonType == soulType).ifPresent(dragonType -> swapWithDragonSoulColor(soulType, THIRD_DRAGON_ROW));
            getDragonType(3).filter(dragonType -> dragonType == soulType).ifPresent(dragonType -> swapWithDragonSoulColor(soulType, FOURTH_DRAGON_ROW));
            SOUL_BAR.forEach(rzKey -> swapWithDragonSoulColor(soulType, SOUL_BAR));
            soulMode = true;
        }
        killedDragonsBar.putAll(getDragonType(0).map(dragonType -> computeDragonColor(dragonType, FIRST_DRAGON_ROW)).orElse(emptyMap()));
        killedDragonsBar.putAll(getDragonType(1).map(dragonType -> computeDragonColor(dragonType, SECOND_DRAGON_ROW)).orElse(emptyMap()));
        killedDragonsBar.putAll(getDragonType(2).map(dragonType -> computeDragonColor(dragonType, THIRD_DRAGON_ROW)).orElse(emptyMap()));
        killedDragonsBar.putAll(getDragonType(3).map(dragonType -> computeDragonColor(dragonType, FOURTH_DRAGON_ROW)).orElse(emptyMap()));
        killedDragonsBar.putAll(dragonSoul.map(dragonType -> computeDragonColor(dragonType, SOUL_BAR)).orElse(emptyMap()));

        return killedDragonsBar;
    }

    private void swapWithDragonSoulColor(DragonType dragonType, List<RzKey> keys) {
        keys.forEach(rzKey -> dragonColorsMap.put(rzKey, dragonType.getSoulColor()));
    }

    private Map<RzKey, Color> computeDragonColor(DragonType dragonType, List<RzKey> keys) {
        Map<RzKey, Color> currentDragonColorMap = new HashMap<>();
        keys.forEach(rzKey -> dragonColorsMap.computeIfAbsent(rzKey, rk -> dragonType.getColor()));
        keys.forEach(rzKey -> currentDragonColorMap.put(rzKey, dragonColorsMap.get(rzKey)));
        return currentDragonColorMap;
    }

    private static Optional<DragonType> getDragonType(int index) {
        final List<DragonType> killedDragons = GameStateHelper.getKilledDragons();
        if (killedDragons.size() > index) {
            return Optional.of(killedDragons.get(index));
        }
        return Optional.empty();
    }

    private static Optional<DragonType> getDragonSoul(List<DragonType> killedDragons) {
        if (killedDragons.size() > 3) {
            return Optional.of(killedDragons.get(3));
        }
        return Optional.empty();
    }
}

File: src/hud/parts/dragons/KilledDragonsBar.ts
import {AnimatedFrame} from '../../../razer-sdk/animation/AnimatedFrame.js';
import {Frame} from '../../../razer-sdk/animation/Frame.js';
import {SimpleFrame} from '../../../razer-sdk/animation/SimpleFrame.js';
import {Color} from '../../../razer-sdk/color/Color.js';
import {RzKey} from '../../../razer-sdk/sdk/RzKey.js';
import {DragonType, dragonTypeGetColor, dragonTypeGetSoulColor} from '../../../rest/eventdata/DragonType.js';
import * as GameStateHelper from '../../../state/GameStateHelper.js';

export const FIRST_DRAGON_ROW: RzKey[] = [RzKey.RZKEY_BACKSLASH];
export const SECOND_DRAGON_ROW: RzKey[] = [RzKey.RZKEY_ENTER];
export const THIRD_DRAGON_ROW: RzKey[] = [RzKey.RZKEY_RSHIFT];
export const FOURTH_DRAGON_ROW: RzKey[] = [RzKey.RZKEY_FN, RzKey.RZKEY_RMENU, RzKey.RZKEY_RCTRL];
export const SOUL_BAR: RzKey[] = [
    RzKey.RZKEY_LCTRL, RzKey.RZKEY_LWIN, RzKey.RZKEY_LALT, RzKey.RZKEY_SPACE, RzKey.RZKEY_RALT,
    RzKey.RZKEY_UP, RzKey.RZKEY_LEFT, RzKey.RZKEY_DOWN, RzKey.RZKEY_RIGHT,
];

const DRAGON_ROWS: RzKey[][] = [FIRST_DRAGON_ROW, SECOND_DRAGON_ROW, THIRD_DRAGON_ROW, FOURTH_DRAGON_ROW];

export class KilledDragonsBar extends AnimatedFrame {
    private soulMode = false;
    private readonly dragonColorsMap = new Map<RzKey, Color>();

    override pollFrame(): Frame {
        this.addAnimationFrame(new SimpleFrame(this.getKilledDragonsBar()));
        return super.pollFrame();
    }

    private getKilledDragonsBar(): Map<RzKey, Color> {
        const killedDragonsBar = new Map<RzKey, Color>();
        const killedDragons = GameStateHelper.getKilledDragons();
        const dragonSoul = getDragonSoul(killedDragons);

        if (dragonSoul !== null && !this.soulMode) {
            for (let i = 0; i < 4; i++) {
                const dt = getDragonType(i);
                if (dt !== null && dt === dragonSoul) {
                    this.swapWithDragonSoulColor(dragonSoul, DRAGON_ROWS[i]);
                }
            }
            this.swapWithDragonSoulColor(dragonSoul, SOUL_BAR);
            this.soulMode = true;
        }

        for (let i = 0; i < 4; i++) {
            const dt = getDragonType(i);
            if (dt !== null) {
                for (const [key, color] of this.computeDragonColor(dt, DRAGON_ROWS[i])) {
                    killedDragonsBar.set(key, color);
                }
            }
        }

        if (dragonSoul !== null) {
            for (const [key, color] of this.computeDragonColor(dragonSoul, SOUL_BAR)) {
                killedDragonsBar.set(key, color);
            }
        }

        return killedDragonsBar;
    }

    private swapWithDragonSoulColor(dragonType: DragonType, keys: RzKey[]): void {
        for (const rzKey of keys) {
            this.dragonColorsMap.set(rzKey, dragonTypeGetSoulColor(dragonType));
        }
    }

    private computeDragonColor(dragonType: DragonType, keys: RzKey[]): Map<RzKey, Color> {
        const currentDragonColorMap = new Map<RzKey, Color>();
        for (const rzKey of keys) {
            if (!this.dragonColorsMap.has(rzKey)) {
                this.dragonColorsMap.set(rzKey, dragonTypeGetColor(dragonType));
            }
        }
        for (const rzKey of keys) {
            currentDragonColorMap.set(rzKey, this.dragonColorsMap.get(rzKey)!);
        }
        return currentDragonColorMap;
    }
}

function getDragonType(index: number): DragonType | null {
    const killedDragons = GameStateHelper.getKilledDragons();
    if (killedDragons.length > index) {
        return killedDragons[index];
    }
    return null;
}

function getDragonSoul(killedDragons: readonly DragonType[]): DragonType | null {
    if (killedDragons.length > 3) {
        return killedDragons[3];
    }
    return null;
}
